SmartDeviceLink/SDLTCPTransport.m

```diff
@@ -52,125 +53,303 @@ - (instancetype)initWithHostName:(NSString *)hostName portNumber:(NSString *)por
+    }
+    if ([self.sendDataQueue count] == 0) {
+        // If send queue is empty, outputStreamHasSpace flag stays in YES. So once sendData is
+        // called, write to the stream will be attempted immediately.
+        return;
     }
 
-    //get socket
-    if ((sock = socket(servinfo->ai_family, servinfo->ai_socktype, servinfo->ai_protocol)) < 0)
-        return (-1);
+    NSMutableData *buffer = [self.sendDataQueue frontBuffer];
+    NSUInteger bufferLen = buffer.length;
+
+    NSInteger bytesWritten = [self.outputStream write:buffer.bytes maxLength:bufferLen];
+    if (bytesWritten < 0) {
+        SDLLogW(@"TCP transport write error: %@", self.outputStream.streamError);
+        [self sdl_onStreamError:self.outputStream];
+        return;
+    } else if (bytesWritten == 0) {
+        SDLLogD(@"TCP transport output stream closed");
+        [self sdl_onStreamEnd:self.outputStream];
+        return;
+    }
 
-    //connect
-    if (connect(sock, servinfo->ai_addr, servinfo->ai_addrlen) < 0) {
-        close(sock);
-        return (-1);
+    if (bytesWritten == bufferLen) {
+        // we have consumed all of data in this buffer
+        [self.sendDataQueue popBuffer];
+    } else {
+        [buffer replaceBytesInRange:NSMakeRange(0, (NSUInteger)bytesWritten) withBytes:NULL length:0];
     }
 
-    freeaddrinfo(servinfo); // free the linked-list
-    return (sock);
+    // the output stream may still have some spaces, but let's wait for another
+    // NSStreamEventHasSpaceAvailable event before writing
+    self.outputStreamHasSpace = NO;
 }
 
-static void TCPCallback(CFSocketRef socket, CFSocketCallBackType type, CFDataRef address, const void *data, void *info) {
-    if (kCFSocketConnectCallBack == type) {
-        SDLTCPTransport *transport = (__bridge SDLTCPTransport *)info;
-        [transport.delegate onTransportConnected];
-    } else if (kCFSocketDataCallBack == type) {
-        SDLTCPTransport *transport = (__bridge SDLTCPTransport *)info;
+- (void)sdl_onConnectionTimedOut:(NSTimer *)timer {
+    SDLLogW(@"TCP connection timed out");
+    [self sdl_cancelIOThread];
 
-        // Check if Core disconnected from us
-        if (CFDataGetLength((CFDataRef)data) <= 0) {
-            SDLLogW(@"Remote system terminated connection, data packet length 0");
-            [transport.delegate onTransportDisconnected];
+    if (!self.transportErrorNotified) {
+        if (!self.transportConnected) { // this condition should be always true
+            SDLLogD(@"Notifying connection timed out error");
+            [self.delegate onError:[NSError sdl_transport_connectionTimedOutError]];
+        } else {
+            [self.delegate onTransportDisconnected];
+        }
+        self.transportErrorNotified = YES;
+    }
+}
 
-            return;
+- (void)sdl_onStreamError:(NSStream *)stream {
+    // stop I/O thread
+    [self sdl_cancelIOThread];
+
+    // avoid notifying multiple error events
+    if (!self.transportErrorNotified) {
```

Comment:
Nothing happens in the `else` case for this `if`. So flip this to `if (self.transportErrorNotified)` and return to keep the rest of the code as left-shifted as possible.